### PR TITLE
Add an error message for "invalid unapply return type" error.

### PR DESCRIPTION
**Pull request by markarasev**
Part of #1589.

### CODE REVIEW

**Comment by maseev** [comment]
> Use the imperative mood in the subject line ("Added" instead of "Add")

I thought I already [fixed](https://github.com/lampepfl/dotty/commit/1278611b7894e0185bf0dfc114f2ce6dab4c9f5e) this. It looks like the dotty-bot hasn't been rebuilt since then :worried:

**Comment by markarasev** [comment]
I fixed the commit message =)

**Comment by senia-psm** [comment]
@markarasev This new feature is called [Name Based Pattern](http://dotty.epfl.ch/docs/reference/changed/pattern-matching.html#name-based-pattern). Please note there there is also a [Product Pattern](http://dotty.epfl.ch/docs/reference/changed/pattern-matching.html#product-pattern).

**Comment by markarasev** [comment]
@senia-psm I updated the explanation to cover these cases.

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/ErrorMessageID.java [hunk 1]
```diff
@@ -113,6 +113,7 @@
     IllegalStartOfStatementID,
     TraitIsExpectedID,
     TraitRedefinedFinalMethodFromAnyRefID,
+    InvalidUnapplyReturnTypeID,
```
Comment (gosubpl) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/ErrorMessageID.java]: Or `UnapplyInvalidReturnTypeID` ? I am now struggling with this for the other one, and decided that `UnapplyInvalidNumberOfArgumentsID` sounds better than `InvalidUnapplyNumberOfArgumentsID`. WDYT?

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 2]
```diff
@@ -92,7 +93,7 @@ object Applications {
     def getTp = extractorMemberType(unapplyResult, nme.get, pos)
 
     def fail = {
-      ctx.error(i"$unapplyResult is not a valid result type of an $unapplyName method of an extractor", pos)
+      ctx.error(InvalidUnapplyReturnType(unapplyResult, unapplyName))
```
Comment (gosubpl) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala]: How is the `pos` being passed here? (sorry for the probably naive question, I am just starting on those error messages)

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/ErrorMessageID.java (hunk 1 above)
Comment (markarasev) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/ErrorMessageID.java]: OK for me, I updated the names.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 2 above)
Comment (markarasev) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala]: I actually misunderstood this value's usage, fixed this.

**Comment by odersky** [comment]
In Dotty, Product subclasses are also legal Unapply result types. So this specific PR is invalid. It would be good to 

 - update the "Changes" section in the references to reflect that, if that is not already the case
 - check whether the error message for types that are still illegal is correct, and update it if necessary.

**Comment by markarasev** [comment]
@odersky do you mean that this part of the explanation is not explicit enough about Product subclasses?
> To be used as an extractor, an unapply method has to return a type that either:
> - [...]
> - is a Product (like a "Tuple2[T1, T2]")

By "changes section in the references" do you mean these docs http://dotty.epfl.ch/docs/reference/changed/pattern-matching.html ? They seem to be up to date.

For your last point, do you mean I should add more unit tests?

**Comment by odersky** [comment]
@OlivierBlanvillain This has been inactive for a year. Let's merge or close this now.

**Pull request merged**